### PR TITLE
Adding consolidateAccount to express

### PR DESCRIPTION
**Pull request by twtaylorbitgo**
Ticket: BG-24544

### CODE REVIEW

File: modules/express/src/clientRoutes.ts [hunk 1]
```diff
@@ -453,6 +453,20 @@ async function handleV2ConsolidateUnspents(req: express.Request) {
   return wallet.consolidateUnspents(req.body);
 }
 
+/**
+ * handle wallet fanout unspents
```
Comment (tylerlevine) [inline modules/express/src/clientRoutes.ts:457]: s/fanout/consolidate/

File: modules/express/src/clientRoutes.ts [hunk 2]
```diff
@@ -453,6 +453,20 @@ async function handleV2ConsolidateUnspents(req: express.Request) {
...
+ * @param req
+ */
+async function handleV2ConsolidateAccount(req: express.Request) {
+  const bitgo = req.bitgo;
+  const coin = bitgo.coin(req.params.coin);
+  if (coin.getFamily() !== 'hbar') {
```
Comment (tylerlevine) [inline modules/express/src/clientRoutes.ts:463]: Please use `CoinFamily.HBAR` from `@bitgo/statics` - these are subject to change.

File: modules/express/src/clientRoutes.ts [hunk 3]
```diff
@@ -453,6 +453,20 @@ async function handleV2ConsolidateUnspents(req: express.Request) {
...
+    throw new Error('invalid coin selected');
+  }
+  const wallet = await coin.wallets().get({ id: req.params.id });
+  return wallet.sendAccountConsolidations(req.body);
```
Comment (tylerlevine) [inline modules/express/src/clientRoutes.ts:467]: Please check wallet return value before use, otherwise you'll emit an unhelpful error to the caller.

**Review by tylerlevine — CHANGES_REQUESTED** [review]
A few comments from me, and this could also use a unit test to make sure behavior remains correct.

File: modules/express/src/clientRoutes.ts (hunk 1 above)
Comment (twtaylorbitgo) [reply modules/express/src/clientRoutes.ts]: Ok done

File: modules/core/example/js/algo/account-consolidation-build.js [hunk 4]
```diff
@@ -56,19 +56,24 @@ Promise.coroutine(function *() {
...
-  // - there is an optional fromAddresses parameter here - you would pass the receive addresses you want to consolidate from
+  // - there is an optional consolidateAddresses parameter here - if you want to pass specific
+  // addresses, pass buildParams above instead of ()
   const consolidationTxes = yield wallet.buildAccountConsolidations();
 
   // this will take all money off receive addresses in the wallet
-  // you can also specify which receive address by passing fromAddresses here:
-  // e.g. { walletPassphrase, fromAddresses: ['onchainReceiveAddress'] }
+  // you can also specify which receive address by passing consolidateAddresses here:
+  // e.g. { walletPassphrase, ΩΩΩ: ['onchainReceiveAddress'] }
```
[the remark this answers is not in the document]
Comment (twtaylorbitgo) [reply modules/core/example/js/algo/account-consolidation-build.js:69]: oh yeah. oops.

**Review by CapnMigraine — APPROVED** [review]
lgtm

File: modules/express/src/clientRoutes.ts [hunk 5]
```diff
@@ -453,6 +454,53 @@ async function handleV2ConsolidateUnspents(req: express.Request) {
...
+
+  if (coin.getFamily() !== CoinFamily.HBAR) {
+    throw new Error('invalid coin selected');
+  }
+
+  const wallet = await coin.wallets().get({ id: req.params.id });
+
+  let result: any;
+  try {
+    result = wallet.sendAccountConsolidations(req.body);
```
Comment (tylerlevine) [inline modules/express/src/clientRoutes.ts:478]: missing `await`?

File: modules/core/src/v2/wallet.ts [hunk 6]
```diff
@@ -63,7 +63,7 @@ export interface Memo {
  * - recipients
  */
 export interface BuildConsolidationTransactionOptions extends PrebuildTransactionOptions {
-  fromAddresses?: string[];
+  consolidateAddresses?: string[];
```
Comment (tylerlevine) [inline modules/core/src/v2/wallet.ts:66]: Was this interface incorrect before? Seems there is no accompanying change in the implementation.

**Review by tylerlevine — CHANGES_REQUESTED** [review]
Looks pretty good, I think it's not going to work quite right though due to a missing await. Was this tested? Can you add a test for this to the express module?

Also, thanks for adding to the examples 👍

File: modules/core/src/v2/wallet.ts (hunk 6 above)
Comment (twtaylorbitgo) [reply modules/core/src/v2/wallet.ts:66]: yup, it was wrong but working if you did the js right

File: modules/express/src/clientRoutes.ts (hunk 5 above)
Comment (twtaylorbitgo) [reply modules/express/src/clientRoutes.ts:478]: oops

**Review by CapnMigraine — APPROVED** [review]

File: modules/express/test/unit/routeHandlers.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+      nock(baseUri)
+        .get(`/api/v2/${coinName}/wallet/${walletId}`)
+        .reply(200, {});
+
+      nock(baseUri)
+        .post(`/api/v2/${coinName}/wallet/${walletId}/consolidateAccount`, { consolidateAddresses })
+        .reply(200, { ttl: 3600, constants: {} });
+
+      const consolidation = handleV2ConsolidateAccount(mockRequest);
```
Comment (tylerlevine) [inline modules/express/test/unit/routeHandlers.ts:77]: Missing `await`

File: modules/express/test/unit/routeHandlers.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+      should.exist(consolidation);
```
Comment (tylerlevine) [inline modules/express/test/unit/routeHandlers.ts:79]: Is there anything else you want to check here except that the return value exists? (note: an unawaited promise will pass this test)

File: modules/express/test/unit/routeHandlers.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    request.agent(app);
+  };
+
+  before(() => {
+    nockServer();
+  });
+
+  describe('Consolidate account', async () => {
+
+    it('should fail on invalid array in body addresses', async () => {
```
Comment (tylerlevine) [inline modules/express/test/unit/routeHandlers.ts:53]: Are there any success tests for this feature?

File: modules/express/test/unit/routeHandlers.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+import * as nock from 'nock';
+import { Environments } from 'bitgo';
+import { handleV2ConsolidateAccount } from '../../src/clientRoutes';
+
+const expressApp = require('../../src/expressApp').app;
+const request = require('supertest-as-promised');
+
+const { BitGo } = require('bitgo');
+
+const oldFetchConstants = BitGo.prototype.fetchConstants;
```
Comment (tylerlevine) [inline modules/express/test/unit/routeHandlers.ts:18]: Could you leave a comment here explaining why we need to monkey patch the bitgo prototype?

File: modules/express/test/unit/routeHandlers.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  BitGo.prototype._constants = undefined;
+
+  return oldFetchConstants.apply(this, arguments);
+};
+
+describe('Bitgo Express', function() {
+
+  const baseUri = Environments['mock'].uri;
+
+  const nockServer = () => {
```
Comment (tylerlevine) [inline modules/express/test/unit/routeHandlers.ts:34]: Is this called somewhere else? If not, why not put it into the `before` block?

**Review by tylerlevine — CHANGES_REQUESTED** [review]
A couple nits on the examples - I won't request changes over that since I know we are under some time pressure to get this feature in, but people hopefully do follow these examples so we should try to make sure they are exemplary.

The missing `await` in the express test should be fixed though.

File: modules/core/example/js/create-wallet-address.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  bitgo.authenticateWithAccessToken({ accessToken: accessToken });
+
+  const wallet = yield bitgo.coin(coin).wallets().get({ id: walletId });
+
+  const newReceiveAddress = yield wallet.createAddress();
+
+  console.log('Wallet ID:', wallet.id());
+  console.log('First Receive Address:', wallet.receiveAddress());
+  console.log('Second Receive Address:', newReceiveAddress);
+})();
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/example/js/create-wallet-address.js]: done

File: modules/express/test/unit/routeHandlers.ts (hunk 7 above)
Comment (barathcj) [reply modules/express/test/unit/routeHandlers.ts]: done

File: modules/core/example/js/algo/account-consolidation-build.js [hunk 13]
```diff
@@ -56,19 +56,24 @@ Promise.coroutine(function *() {
...
     // this is one example of how you might send only the first consolidation from this group
     const unsignedConsolidation = consolidationTxes[0];
-    const sendConsolidations = yield wallet.sendAccountConsolidations({ walletPassphrase, prebuildTx: unsignedConsolidation });
+    const sendConsolidations = yield wallet.sendAccountConsolidation({ walletPassphrase, prebuildTx: unsignedConsolidation });
     console.dir(sendConsolidations, { depth: 6 });
   } catch (e) {
     console.error(e);
   }
-});
+})();
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/example/js/algo/account-consolidation-build.js]: done

File: modules/core/example/js/algo/account-consolidation-build.js [hunk 14]
```diff
@@ -56,19 +56,24 @@ Promise.coroutine(function *() {
...
-  // you can also specify which receive address by passing fromAddresses here:
-  // e.g. { walletPassphrase, fromAddresses: ['onchainReceiveAddress'] }
+  // you can also specify which receive address by passing consolidateAddresses here:
+  // e.g. { walletPassphrase, consolidateAddresses: ['onchainReceiveAddress'] }
+
   try {
     // this is one example of how you might send only the first consolidation from this group
     const unsignedConsolidation = consolidationTxes[0];
-    const sendConsolidations = yield wallet.sendAccountConsolidations({ walletPassphrase, prebuildTx: unsignedConsolidation });
+    const sendConsolidations = yield wallet.sendAccountConsolidation({ walletPassphrase, prebuildTx: unsignedConsolidation });
```
[the remark this answers is not in the document]
Comment (barathcj) [reply modules/core/example/js/algo/account-consolidation-build.js:74]: I think it should be sendAccountConsolidation. 
sendAccountConsolidations does something else, maybe @twtaylorbitgo can comment

**Comment by twtaylorbitgo** [comment]
I just tested this in prod, the consolidate addresses singularly and broad sweep functionality works in express.

**Review by CapnMigraine — APPROVED** [review]
lgtm

**Review by tylerlevine — APPROVED** [review]

**Pull request merged**